src/dune_engine/build_system.ml

```diff
@@ -947,6 +953,35 @@ end = struct
       Path.Build.Set.to_list source_files_to_ignore
       |> Path.Source.Set.of_list_map ~f:Path.Build.drop_build_context_exn
     in
+
+    (* If a [.merlin] file is present in the [Promoted_to_delete] set but not in
+       the [Source_files_to_ignore] that means the rule that ordered its
+       promotion is no more valid. This would happen when upgrading to Dune 2.8
+       from ealier version without and building uncleaned projects. We delete
+       these leftover files here. *)
+    let merlin_file = ".merlin" in
+    let source_dir = Path.Build.drop_build_context_exn dir in
+    let merlin_in_src = Path.Source.(relative source_dir merlin_file) in
+    let source_files_to_ignore =
+      if
+        Path.Set.mem (Promoted_to_delete.load ()) (Path.source merlin_in_src)
```

Comment:
Calling `load` once per directory sounds expensive. My understanding is that we expect this file to have size O(n) where n is the number of directories with .merlin files (eve without taking into account anything else the user happens to promote) and you're loading it n times, which makes for O(n^2) time complexity.

I think we should be able to change it so that `load` is only called once, and `Promoted_to_delete.dump` simply writes out the db without calling `load` again.